masonry/src/passes/update.rs

```diff
@@ -145,3 +146,59 @@ pub(crate) fn run_update_pointer_pass(root: &mut RenderRoot, root_state: &mut Wi
     // Pass root widget state to synthetic state create at beginning of pass
     root_state.merge_up(root.widget_arena.get_state_mut(root.root.id()).item);
 }
+
+// ----------------
+
+fn update_disabled_for_widget(
+    global_state: &mut RenderRootState,
+    mut widget: ArenaMut<'_, Box<dyn Widget>>,
+    mut state: ArenaMut<'_, WidgetState>,
+    parent_disabled: bool,
+) {
+    let _span = widget.item.make_trace_span().entered();
+    let id = state.item.id;
+
+    let disabled = state.item.is_explicitly_disabled | parent_disabled;
+    if !state.item.needs_update_disabled && disabled == state.item.is_disabled {
+        return;
+    }
+
+    if disabled != state.item.is_disabled {
+        let mut ctx = LifeCycleCtx {
+            global_state,
+            widget_state: state.item,
+            widget_state_children: state.children.reborrow_mut(),
+            widget_children: widget.children.reborrow_mut(),
+        };
+        widget
+            .item
+            .lifecycle(&mut ctx, &LifeCycle::DisabledChanged(disabled));
+        state.item.is_disabled = disabled;
+    }
+
+    state.item.needs_update_disabled = false;
+
+    if disabled && global_state.next_focused_widget == Some(id) {
+        // This may get overwritten. That's ok, because either way the
+        // focused widget, if there's one, won't be disabled.
+        global_state.next_focused_widget = None;
```

Comment:
If a widget is forcefully disabled, I'd expect pressing <kbd>Tab</kbd> next to focus what the next widget would be, rather than restarting at the beginning. It's hard to get a feel for this, though - I think this behaviour is OK at the moment, as it will change when we do more careful focus handling.